### PR TITLE
Added support of running Helmsman on Windows

### PR DESCRIPTION
**Pull request by PeterBrun**
Any execution of a command using bash with the command parameter has been replaced with just the expected command executable (helm and kubectl). And instead of building the arguments as a single string, they are now built using string arrays. This solves the "quoting" problem, that is there when using a single string with all arguments. For example, before this change, a path had to be quoted on Windows to work. By just building string arrays with all the arguments, this problem is solved by Go' os/exec package.

It also solves the problem of using a password protected Helm repo, where the username contains a '$' (for example Harbor creates robot accounts with this). This could be escaped using '$$', but then it still was being interpreted by bash as an environment variable. This could have been solved by adding quotes around in the old solution, but it is just simpler to just use the "native" string array arguments approach as described above. And still if the environment variable was intended as input, then it is already applied by the "substituteEnv" functionality.

### CODE REVIEW

File: decision_maker.go [hunk 1]
```diff
@@ -212,12 +216,14 @@ func inspectUpgradeScenario(r *release, rs releaseState) {
 
 	if r.Namespace == rs.Namespace {
 
-		version, msg := getChartVersion(r)
-		if msg != "" {
-			logError(msg)
-			return
+		if !isLocalChart(r.Chart) {
```
Comment (nvanheuverzwijn) [inline decision_maker.go]: This part of the patch should go into another PR and be merged as soon as possible.

Comment (psbtrifork) [reply decision_maker.go]: I have removed this patch from this PR. Awaiting PR #291 instead.

Comment (psbtrifork) [reply decision_maker.go]: I have removed this patch from this PR. Awaiting PR #291 instead.

Comment (PeterBrun) [reply decision_maker.go]: I have removed this patch from this PR. Awaiting PR #291 instead.

**Comment by nvanheuverzwijn** [comment]
@luisdavim any reason why commands used to be call with `bash -c` instead of using `helm` directly ? This PR changes how helm is called, not sure if it has any impact or not.

File: decision_maker.go [hunk 2]
```diff
@@ -393,36 +397,37 @@ func getValuesFiles(r *release) string {
...
 }
 
 // getSetValues returns --set params to be used with helm install/upgrade commands
-func getSetValues(r *release) string {
-	result := ""
+func getSetValues(r *release) []string {
+	result := []string{}
 	for k, v := range r.Set {
-		result = result + " --set " + k + "=\"" + strings.Replace(v, ",", "\\,", -1) + "\""
+		result = append(result, "--set", k+"="+strings.Replace(v, ",", "\\,", -1)+"")
```
Comment (nvanheuverzwijn) [inline decision_maker.go]: This here seems to be different. The resulting string would be `k=something` instead of k="something". It would break if `v` have any space. Unless `v` is never supposed to have space in it.

File: decision_maker.go [hunk 3]
```diff
@@ -393,36 +397,37 @@ func getValuesFiles(r *release) string {
...
 }
 
 // getSetStringValues returns --set-string params to be used with helm install/upgrade commands
-func getSetStringValues(r *release) string {
-	result := ""
+func getSetStringValues(r *release) []string {
+	result := []string{}
 	for k, v := range r.SetString {
-		result = result + " --set-string " + k + "=\"" + strings.Replace(v, ",", "\\,", -1) + "\""
+		result = append(result, "--set-string", k+"="+strings.Replace(v, ",", "\\,", -1)+"")
```
Comment (nvanheuverzwijn) [inline decision_maker.go]: Same here.

File: helm_helpers.go [hunk 4]
```diff
@@ -254,12 +255,11 @@ func getReleaseChartVersion(rs releaseState) string {
...
 // getNSTLSFlags returns TLS flags for a given namespace if it's deployed with TLS
-func getNSTLSFlags(namespace string) string {
-	tls := ""
+func getNSTLSFlags(namespace string) []string {
+	tls := []string{}
 	ns := s.Namespaces[namespace]
 	if tillerTLSEnabled(ns) {
-
-		tls = " --tls --tls-ca-cert " + namespace + "-ca.cert --tls-cert " + namespace + "-client.cert --tls-key " + namespace + "-client.key "
+		tls = append(tls, "--tls", "--tls-ca-cert", namespace+"-ca.cert", "--tls-cert", namespace+"-client.cert", "--tls-key", namespace+"-client.key")
```
Comment (nvanheuverzwijn) [inline helm_helpers.go]: For consistency, I would add space between `+`

File: helm_helpers.go [hunk 5]
```diff
@@ -325,8 +325,8 @@ func getChartVersion(r *release) (string, string) {
 		return r.Version, ""
 	}
 	cmd := command{
-		Cmd:         "bash",
-		Args:        []string{"-c", "helm search " + r.Chart + " --version " + strconv.Quote(r.Version)},
+		Cmd:         "helm",
+		Args:        []string{"search", r.Chart, "--version", r.Version},
```
Comment (nvanheuverzwijn) [inline helm_helpers.go]: Here, `strconv.Quote` is removed. I am not too sure what the outcome of doing this is.

File: helm_helpers.go [hunk 6]
```diff
@@ -395,13 +395,13 @@ func addHelmRepos(repos map[string]string) (bool, string) {
...
-			basicAuth = " --username " + u.User.Username() + " --password " + p
+			basicAuthArgs = append(basicAuthArgs, "--username", u.User.Username(), "--password", p)
 
 		}
 
 		cmd := command{
-			Cmd:         "bash",
-			Args:        []string{"-c", "helm repo add " + basicAuth + " " + repoName + " " + strconv.Quote(repoLink)},
+			Cmd:         "helm",
+			Args:        concat([]string{"repo", "add", repoName, repoLink}, basicAuthArgs),
```
Comment (nvanheuverzwijn) [inline helm_helpers.go]: Same here, `strconv.Quote` removed.

File: helm_helpers.go [hunk 7]
```diff
@@ -506,8 +506,8 @@ func deployTiller(namespace string, serviceAccount string, defaultServiceAccount
 // initHelmClientOnly initializes the helm client only (without deploying Tiller)
 func initHelmClientOnly() (bool, string) {
 	cmd := command{
-		Cmd:         "bash",
-		Args:        []string{"-c", "helm init --client-only "},
+		Cmd:         "helm",
+		Args:        []string{"init", "--client-only"},
```
Comment (nvanheuverzwijn) [inline helm_helpers.go]: There used to be a space after `--client-only` (`"--client-only"` instead of `"--client-only "`). Not sure what the impact is, might be a typo that this PR fixes now.

File: helm_helpers_test.go [hunk 8]
```diff
@@ -8,11 +8,11 @@ import (
...
 	t.Log("setup test case")
-	os.MkdirAll("/tmp/helmsman-tests/myapp", os.ModePerm)
-	os.MkdirAll("/tmp/helmsman-tests/dir-with space/myapp", os.ModePerm)
+	os.MkdirAll(os.TempDir()+"/helmsman-tests/myapp", os.ModePerm)
+	os.MkdirAll(os.TempDir()+"/helmsman-tests/dir-with space/myapp", os.ModePerm)
 	cmd := command{
-		Cmd:         "bash",
-		Args:        []string{"-c", "helm create  '/tmp/helmsman-tests/dir-with space/myapp'"},
+		Cmd:         "helm",
+		Args:        []string{"create", os.TempDir() + "/helmsman-tests/dir-with space/myapp"},
```
Comment (nvanheuverzwijn) [inline helm_helpers_test.go]: Single quote `'` are removed here.

File: helm_helpers_test.go [hunk 9]
```diff
@@ -138,7 +138,7 @@ func Test_validateReleaseCharts(t *testing.T) {
 						Namespace:       "",
 						Enabled:         true,
 						Chart:           "stable/prometheus",
-						Version:         "",
+						Version:         "*",
```
Comment (nvanheuverzwijn) [inline helm_helpers_test.go]: Why change `Version` from `""` to `"*"` ? Did this PR breaks this test ?

File: helm_helpers_test.go [hunk 10]
```diff
@@ -233,7 +233,7 @@ func Test_validateReleaseCharts(t *testing.T) {
 				targetMap[target] = true
 			}
 			if got, msg := validateReleaseCharts(tt.args.apps); got != tt.want {
-				t.Errorf("getReleaseChartName() = %v, want %v , msg: %v", got, tt.want, msg)
+				t.Errorf("validateReleaseCharts() = %v, want %v , msg: %v", got, tt.want, msg)
```
Comment (nvanheuverzwijn) [inline helm_helpers_test.go]: Does this change have anything to do with windows ? Is it because the test did not report the correct behavior ?

File: helm_helpers_test.go [hunk 11]
```diff
@@ -334,7 +334,7 @@ func Test_getReleaseChartVersion(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
 			t.Log(tt.want)
 			if got := getReleaseChartVersion(tt.args.r); got != tt.want {
-				t.Errorf("getReleaseChartName() = %v, want %v", got, tt.want)
+				t.Errorf("getReleaseChartVersion() = %v, want %v", got, tt.want)
```
Comment (nvanheuverzwijn) [inline helm_helpers_test.go]: Does this change have anything to do with windows ? Is it because the test did not report the correct behavior ?

File: kube_helpers.go [hunk 12]
```diff
@@ -230,23 +230,23 @@ func createContext() (bool, string) {
...
-		setCredentialsCmd = "kubectl config set-credentials " + s.Settings.Username + " --token=" + token
+		setCredentialsCmdArgs = append(setCredentialsCmdArgs, "config", "set-credentials", s.Settings.Username, "--token="+token)
 	} else {
-		setCredentialsCmd = "kubectl config set-credentials " + s.Settings.Username + " --username=" + s.Settings.Username +
-			" --password=" + s.Settings.Password + " --client-key=" + caKey
+		setCredentialsCmdArgs = append(setCredentialsCmdArgs, "config", "set-credentials", s.Settings.Username, "--username="+s.Settings.Username,
+			"--password="+s.Settings.Password, "--client-key="+caKey)
 		if caClient != "" {
-			setCredentialsCmd = setCredentialsCmd + " --client-certificate=" + caClient
+			setCredentialsCmdArgs = append(setCredentialsCmdArgs, "--client-certificate="+caClient)
```
Comment (nvanheuverzwijn) [inline kube_helpers.go]: I would add space between `+` for consistency.

File: decision_maker.go (hunk 2 above)
Comment (PeterBrun) [reply decision_maker.go]: No, it does not break if you enter a space.

File: decision_maker.go (hunk 3 above)
Comment (PeterBrun) [reply decision_maker.go]: No, it does not break if you enter a space.

File: decision_maker.go (hunk 2 above)
Comment (nvanheuverzwijn) [reply decision_maker.go]: Gotcha, just making sure.

File: helm_helpers.go (hunk 4 above)
Comment (PeterBrun) [reply helm_helpers.go]: That is how gofmt formats that kind of line. There is over 100 examples of this kind in the existing codebase.

File: helm_helpers.go (hunk 5 above)
Comment (PeterBrun) [reply helm_helpers.go]: Please read my description:

> This solves the "quoting" problem, that is there when using a single string with all arguments.

File: helm_helpers.go (hunk 6 above)
Comment (PeterBrun) [reply helm_helpers.go]: Please read my description:

> This solves the "quoting" problem, that is there when using a single string with all arguments.

File: helm_helpers_test.go (hunk 8 above)
Comment (PeterBrun) [reply helm_helpers_test.go]: That is because the 'qoute' hack is no longer needed.

File: helm_helpers_test.go (hunk 9 above)
Comment (PeterBrun) [reply helm_helpers_test.go]: I will update the "helm search" code to check for the empty string.

File: helm_helpers_test.go (hunk 10 above)
Comment (PeterBrun) [reply helm_helpers_test.go]: I think the old code is wrong. Would you agree? Especially when debugging a unit test, it says something that is not tested here. Here "validateReleaseCharts" and now it says "validateReleaseCharts" failed if it fails.

File: helm_helpers_test.go (hunk 11 above)
Comment (PeterBrun) [reply helm_helpers_test.go]: I think the old code is wrong. Would you agree? Especially when debugging a unit test, it says something that is not tested here. Here "getReleaseChartVersion" and now it says "getReleaseChartVersion" failed if it fails.

File: kube_helpers.go (hunk 12 above)
Comment (PeterBrun) [reply kube_helpers.go]: That is how gofmt formats that kind of line. There is over 120 examples of this kind in the existing code base.

File: decision_maker.go (hunk 2 above)
Comment (nvanheuverzwijn) [reply decision_maker.go]: Refactors are hard and sometimes, small details will destroy the effort, just making sure that this particular ommision of the double quote was really intended and not a typo.

**Comment by nvanheuverzwijn** [comment]
@PeterBrun I have highlighted details that seemed to me that would change the behavior but it might be OK. I just wanted to highlight them for you to make sure that everything is intended by this patch and not a typo or a bug in a search & replace.

**Comment by PeterBrun** [comment]
@nvanheuverzwijn 👍

@luisdavim I have mostly tested on Windows. Using it there with some clusters. I have only tested the Linux build using the provided Linux tests (not with a real cluster).

**Comment by luisdavim** [comment]
@sami-alajrami this LGTM but I wonder if there was any particular reason to start the helm and kubectl commands in a sub-shell instead of calling them directly.

**Comment by luisdavim** [comment]
@mkubaczyk do you know what a sub-shell is being used?

**Comment by luisdavim** [comment]
Hi can you take a look at the conflicts? Thanks.

**Comment by sami-alajrami** [comment]
> @sami-alajrami this LGTM but I wonder if there was any particular reason to start the helm and kubectl commands in a sub-shell instead of calling them directly.

No, I don't think there was a special reason. Unless if it was to get error codes back .. but my memory is vague on it

**Comment by luisdavim** [comment]
I that case I think we can merge this as soon as the conflicts are addressed. Thanks
